### PR TITLE
webframe: api to register scheme as privileged

### LINKED ISSUE #2831: ServiceWorker does not allow 'file://' origin
I am trying to intercept image load request from my HTML to https requests. Specifically my app has a URL "images/test.png" in the HTML and electron tries to load that with a file scheme  "file:///images/test.png". I would like to rewrite this request to https://mydomain.com/images/test.png

For this I was trying to use a service worker since they allow to intercept the network requests and modify the URL.

The problem is that electron (well chromium) does not allow "file://" origin. Chromium states that Service Workers require a secure origin, such as HTTPS. 

The exact error is:
"Failed to register a ServiceWorker: The URL protocol of the current origin ('file://') is not supported."

Is there a way to get around that with electron?

A somewhat related Chromium submit is https://codereview.chromium.org/1191793003/#ps40001

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #2831

### CODE REVIEW

**Comment by deepak1556** [comment]
Although this fixes the security error for registering service workers, with current version chromium there are hard coded checks for scheme is http/https to handle service worker  requests, hence they wont work. But this has been fixed in this [CL](https://codereview.chromium.org/1192013003) which gives embedders's a chance to register schemes that can handle service workers, we need to wait for that.

File: atom/renderer/api/atom_api_web_frame.cc [hunk 1]
```diff
@@ -86,16 +86,18 @@ void WebFrame::SetSpellCheckProvider(mate::Arguments* args,
   web_frame_->view()->setSpellCheckClient(spell_check_client_.get());
 }
 
-void WebFrame::RegisterURLSchemeAsSecure(const std::string& scheme) {
```
Comment (zcbenz) [inline atom/renderer/api/atom_api_web_frame.cc]: We still need to keep it for backward compatibility.

File: atom/renderer/api/atom_api_web_frame.cc [hunk 2]
```diff
@@ -86,16 +86,18 @@ void WebFrame::SetSpellCheckProvider(mate::Arguments* args,
...
+    blink::WebSecurityPolicy::registerURLSchemeAsBypassingContentSecurityPolicy,
+    blink::WebSecurityPolicy::registerURLSchemeAsAllowingServiceWorkers
+  };
 
-void WebFrame::RegisterURLSchemeAsBypassingCsp(const std::string& scheme) {
-  // Register scheme to bypass pages's Content Security Policy.
-  blink::WebSecurityPolicy::registerURLSchemeAsBypassingContentSecurityPolicy(
-      blink::WebString::fromUTF8(scheme));
+  blink::WebString privileged_scheme(blink::WebString::fromUTF8(scheme));
+  for (RegisterFunction func : register_functions)
```
Comment (zcbenz) [inline atom/renderer/api/atom_api_web_frame.cc]: I don't see a reason to call the APIs like this, we still need to enumerate every function and the code doesn't look cleaner.

Comment (deepak1556) [inline atom/renderer/api/atom_api_web_frame.cc]: Fixed, thanks!

**Comment by zcbenz** [comment]
Thanks!

**Comment by rogeru** [comment]
Thank you. Do we know when the missing Chrome issue will be available to be included in Electron?

**Comment by deepak1556** [comment]
@rogeru that particular CLs' parent issue is set for M46 release, not sure when. You could open an issue for tracking, will be implemented once the api becomes available.

**Comment by rogeru** [comment]
Thanks @deepak1556.

I am using registerUrlSchemeAsPrivileged which gets me passed the previous error about file:// not being allowed. The next error is "SecurityError: Failed to register a ServiceWorker: No URL is associated with the caller's document."
It seems we also need to wait for https://code.google.com/p/chromium/issues/detail?id=522900.

**Comment by deepak1556** [comment]
Yes thats because of the checks i was talking about which only does `SchemeIsHTTPorHTTPS` everywhere now. Once the upgrade its replaced by generic [OriginCanAccessServiceWorkers](https://chromium.googlesource.com/chromium/src/+/1db36fd%5E%21/#F17) check everywhere. This maintains a map of schemes allowed which is obtained from `ContentClient::AddServiceWorkerSchemes`

**Comment by havenchyk** [comment]
@deepak1556 is this fix already in the release? Can I already use it?

**Comment by deepak1556** [comment]
@havenchyk nope need to wait for chromium upgrade, can be tracked here https://github.com/atom/electron/issues/3312#issuecomment-153632338

**Comment by havenchyk** [comment]
@deepak1556 how about now? I guess there was an [update to Chrome 47](https://github.com/atom/electron/pull/3714)?

**Comment by deepak1556** [comment]
@havenchyk yup need to wait on the PR.

**Pull request merged**